controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,561 @@
  */
 package io.pravega.controller.store.stream;
 
+import com.google.common.base.Preconditions;
+import io.pravega.client.stream.StreamConfiguration;
 import io.pravega.common.concurrent.FutureHelpers;
+import io.pravega.common.util.BitConverter;
 import io.pravega.controller.store.stream.tables.ActiveTxnRecord;
+import io.pravega.controller.store.stream.tables.CompletedTxnRecord;
+import io.pravega.controller.store.stream.tables.Create;
+import io.pravega.controller.store.stream.tables.Data;
 import io.pravega.controller.store.stream.tables.State;
-import io.pravega.client.stream.StreamConfiguration;
-import com.google.common.base.Preconditions;
-import lombok.AllArgsConstructor;
-import org.apache.commons.lang.NotImplementedException;
+import io.pravega.controller.store.stream.tables.TableHelper;
+import org.apache.commons.lang.SerializationUtils;
 
-import java.util.AbstractMap.SimpleEntry;
-import java.util.ArrayList;
+import javax.annotation.concurrent.GuardedBy;
+import java.util.AbstractMap;
+import java.util.Arrays;
 import java.util.Collections;
-import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
+import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
-/**
- * Stream properties
- * <p>
- * This class is no longer consistent and mostly not Implemented. Deprecating it.
- */
-class InMemoryStream implements Stream {
-    private final String streamName;
-    private final String scopeName;
-    private StreamConfiguration configuration;
-    private State state;
+public class InMemoryStream extends PersistentStreamBase<Integer> {
 
-    /**
-     * Stores all segments in the stream, ordered by number, which implies that
-     * these segments are also ordered in the increasing order of their start times.
-     * Segment number is the index of that segment in this list.
-     */
-    private final List<InMemorySegment> segments = new ArrayList<>();
+    private final AtomicLong creationTime = new AtomicLong(Long.MIN_VALUE);
+    private final Object lock = new Object();
 
+    @GuardedBy("lock")
+    private StreamConfiguration configuration;
+    @GuardedBy("lock")
+    private volatile Data<Integer> state;
```

Comment:
i had used lock and 'volatile' in combination because i did not want to take a lock for reads. Making it volatile meant it would be read from the main memory if there was any update. 
However, i have now removed volatile and am doing reads under lock too.